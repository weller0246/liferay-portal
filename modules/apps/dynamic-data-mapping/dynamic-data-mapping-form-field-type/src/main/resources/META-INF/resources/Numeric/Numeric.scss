.ddm-form-field-type__numeric {
	&--rtl {
		text-align: right;
	}
}
