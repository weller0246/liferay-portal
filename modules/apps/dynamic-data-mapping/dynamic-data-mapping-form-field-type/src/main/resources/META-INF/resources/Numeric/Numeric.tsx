/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

import {ClayInput} from '@clayui/form';

// @ts-ignore

import {useFormState} from 'data-engine-js-components-web';
import React, {ChangeEventHandler, FocusEventHandler, useMemo} from 'react';
import createNumberMask from 'text-mask-addons/dist/createNumberMask';

// @ts-ignore

import {conformToMask} from 'vanilla-text-mask';

// @ts-ignore

import {FieldBase} from '../FieldBase/ReactFieldBase.es';
import {ISymbols} from '../NumericInputMask/NumericInputMask';

// @ts-ignore

import withConfirmationField from '../util/withConfirmationField.es';

const adaptiveMask = (rawValue: string, inputMaskFormat: string) => {
	const generateMask = (mask: string): string => {
		if (!mask.includes('0')) {
			return mask;
		}

		const inputNumbers = rawValue.match(/\d/g)?.length ?? 0;
		const mandatorySize = mask.match(/9/g)?.length ?? 0;
		if (inputNumbers <= mandatorySize) {
			return mask.replace(/0/g, '');
		}

		return generateMask(mask.replace('0', '9'));
	};

	return [...generateMask(inputMaskFormat)].map((char) =>
		char === '9' ? /\d/ : char
	);
};

const getMaskedValue = ({
	append,
	appendType,
	dataType,
	inputMask,
	inputMaskFormat,
	symbols,
	value,
}: {
	append: string;
	appendType?: 'prefix' | 'suffix';
	dataType: NumericDataType;
	inputMask?: boolean;
	inputMaskFormat?: string;
	symbols: ISymbols;
	value: string;
}): IMaskedNumber => {
	let mask;

	if (inputMask) {
		if (dataType === 'double') {
			const config: INumberMaskConfig = {
				allowDecimal: true,
				allowLeadingZeroes: true,
				allowNegative: true,
				decimalSymbol: symbols.decimalSymbol,
				includeThousandsSeparator: Boolean(symbols.thousandsSeparator),
				prefix: appendType === 'prefix' ? append : '',
				suffix: appendType === 'suffix' ? append : '',
				thousandsSeparatorSymbol: symbols.thousandsSeparator,
			};

			mask = createNumberMask(config);
		}
		else {
			mask = adaptiveMask(value, inputMaskFormat as string);
		}
	}
	else {
		const config: INumberMaskConfig = {
			allowLeadingZeroes: true,
			allowNegative: true,
			includeThousandsSeparator: false,
			prefix: '',
		};

		if (dataType === 'double') {
			config.allowDecimal = true;
			config.decimalLimit = null;
			config.decimalSymbol = symbols.decimalSymbol;
		}
		mask = createNumberMask(config);

		if (typeof value === 'string') {
			if (!value) {
				return {masked: '', raw: ''};
			}
			value = value.replace(symbols.decimalSymbol, '.');
			if (dataType == 'integer' && value.includes('.')) {
				value = value.replace(symbols.decimalSymbol, '');
			}
		}
		value = value.replace('.', symbols.decimalSymbol);
	}

	const {conformedValue: masked} = conformToMask(value, mask, {
		guide: false,
		keepCharPositions: false,
		placeholderChar: '\u2000',
	});
	const regex = new RegExp(`[^${symbols.decimalSymbol}|\\d]`, 'g');

	return {masked, raw: inputMask ? masked.replace(regex, '') : masked};
};

const Numeric: React.FC<IProps> = ({
	append,
	appendType,
	dataType = 'integer',
	defaultLanguageId,
	id,
	inputMask,
	inputMaskFormat,
	localizedValue,
	name,
	onBlur,
	onChange,
	onFocus,
	placeholder,
	predefinedValue,
	readOnly,
	symbols: symbolsProp = {decimalSymbol: '.'},
	value,
	...otherProps
}) => {
	const {editingLanguageId} = useFormState();
	const symbols = useMemo<ISymbols>(() => {
		return inputMask
			? {
					decimalSymbol: symbolsProp.decimalSymbol,
					thousandsSeparator:
						symbolsProp.thousandsSeparator == 'none'
							? null
							: symbolsProp.thousandsSeparator,
			  }
			: symbolsProp;
	}, [inputMask, symbolsProp]);

	const inputValue = useMemo<IMaskedNumber>(() => {
		const newValue =
			((localizedValue?.[editingLanguageId] ??
				localizedValue?.[defaultLanguageId]) ||
				value) ??
			predefinedValue ??
			'';

		return getMaskedValue({
			append,
			appendType,
			dataType,
			inputMask,
			inputMaskFormat,
			symbols,
			value: newValue,
		});
	}, [
		append,
		appendType,
		dataType,
		symbols,
		defaultLanguageId,
		editingLanguageId,
		inputMask,
		inputMaskFormat,
		localizedValue,
		predefinedValue,
		value,
	]);

	const handleChange: ChangeEventHandler<HTMLInputElement> = ({
		target: {value},
	}) => {
		const rawValue = value.replace(/\D/g, '');
		if (
			(inputValue.masked?.length ?? 0) - value.length === 1 &&
			(inputValue.raw?.length ?? 0) === rawValue.length
		) {
			value = inputValue.raw.slice(0, -1);
		}

		const newValue = getMaskedValue({
			append,
			appendType,
			dataType,
			inputMask,
			inputMaskFormat,
			symbols,
			value,
		});
		if (newValue.masked !== inputValue.masked) {
			onChange({target: {value: newValue.raw}});
		}
	};

	return (
		<FieldBase
			{...otherProps}
			id={id}
			localizedValue={localizedValue}
			name={name}
			readOnly={readOnly}
		>
			<ClayInput

				// @ts-ignore

				dir={Liferay.Language.direction[editingLanguageId]}
				disabled={readOnly}
				id={id}
				lang={editingLanguageId}
				name={`${name}${inputMask ? '_masked' : ''}`}
				onBlur={onBlur}
				onChange={handleChange}
				onFocus={onFocus}
				placeholder={
					placeholder ||
					(inputMask
						? inputMaskFormat?.replace(/\d/g, '_')
						: undefined)
				}
				type="text"
				value={inputValue.masked}
			/>
			{inputMask && (
				<input name={name} type="hidden" value={inputValue.raw} />
			)}
		</FieldBase>
	);
};

export {Numeric};
export default withConfirmationField(Numeric);

interface IMaskedNumber {
	masked: string;
	raw: string;
}

interface INumberMaskConfig {
	allowDecimal?: boolean;
	allowLeadingZeroes: boolean;
	allowNegative: boolean;
	decimalLimit?: number | null;
	decimalSymbol?: string;
	includeThousandsSeparator: boolean;
	prefix: string;
	suffix?: string;
	thousandsSeparatorSymbol?: string | null;
}
interface IProps {
	append: string;
	appendType: 'prefix' | 'suffix';
	dataType: NumericDataType;
	defaultLanguageId: string;
	id: string;
	inputMask?: boolean;
	inputMaskFormat?: string;
	localizedValue?: {[key: string]: string};
	name: string;
	onBlur: FocusEventHandler<HTMLInputElement>;
	onChange: (event: {target: {value: string}}) => void;
	onFocus: FocusEventHandler<HTMLInputElement>;
	placeholder?: string;
	predefinedValue?: string;
	readOnly: boolean;
	symbols: ISymbols;
	thousandsSeparator?: [',' | '.' | ' ' | "'" | 'none'];
	value?: string;
}

type NumericDataType = 'integer' | 'double';
