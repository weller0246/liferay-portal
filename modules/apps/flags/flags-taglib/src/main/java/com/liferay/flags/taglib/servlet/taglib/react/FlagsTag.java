/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

package com.liferay.flags.taglib.servlet.taglib.react;

import com.liferay.flags.taglib.internal.servlet.ServletContextUtil;
import com.liferay.flags.taglib.internal.servlet.taglib.util.FlagsTagUtil;
import com.liferay.portal.kernel.exception.PortalException;
import com.liferay.portal.kernel.json.JSONObject;
import com.liferay.portal.kernel.json.JSONUtil;
import com.liferay.portal.kernel.language.LanguageUtil;
import com.liferay.portal.kernel.log.Log;
import com.liferay.portal.kernel.log.LogFactoryUtil;
import com.liferay.portal.kernel.model.Company;
import com.liferay.portal.kernel.model.User;
import com.liferay.portal.kernel.theme.ThemeDisplay;
import com.liferay.portal.kernel.util.AggregateResourceBundle;
import com.liferay.portal.kernel.util.HashMapBuilder;
import com.liferay.portal.kernel.util.PortalUtil;
import com.liferay.portal.kernel.util.PortletKeys;
import com.liferay.portal.kernel.util.Validator;
import com.liferay.portal.kernel.util.WebKeys;
import com.liferay.portal.language.LanguageResources;
import com.liferay.taglib.util.IncludeTag;
import com.liferay.taglib.util.TagResourceBundleUtil;

import java.util.Map;
import java.util.ResourceBundle;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.jsp.PageContext;

/**
 * @author Ambrín Chaudhary
 * @author Jorge González
 */
public class FlagsTag extends IncludeTag {

	public String getClassName() {
		return _className;
	}

	public long getClassPK() {
		return _classPK;
	}

	public String getContentTitle() {
		return _contentTitle;
	}

	public String getContentURL() {
		return _contentURL;
	}

	public String getElementClasses() {
		return _elementClasses;
	}

	public String getMessage() {
		return _message;
	}

	public long getReportedUserId() {
		return _reportedUserId;
	}

	public boolean isEnabled() {
		return _enabled;
	}

	public boolean isLabel() {
		return _label;
	}

	public void setClassName(String className) {
		_className = className;
	}

	public void setClassPK(long classPK) {
		_classPK = classPK;
	}

	public void setContentTitle(String contentTitle) {
		_contentTitle = contentTitle;
	}

	public void setContentURL(String contentURL) {
		_contentURL = contentURL;
	}

	public void setElementClasses(String elementClasses) {
		_elementClasses = elementClasses;
	}

	public void setEnabled(boolean enabled) {
		_enabled = enabled;
	}

	public void setLabel(boolean label) {
		_label = label;
	}

	public void setMessage(String message) {
		_message = message;
	}

	@Override
	public void setPageContext(PageContext pageContext) {
		super.setPageContext(pageContext);

		setServletContext(ServletContextUtil.getServletContext());
	}

	public void setReportedUserId(long reportedUserId) {
		_reportedUserId = reportedUserId;
	}

	@Override
	protected void cleanUp() {
		super.cleanUp();

		_className = null;
		_classPK = 0;
		_contentTitle = null;
		_contentURL = null;
		_elementClasses = null;
		_enabled = true;
		_label = true;
		_message = null;
		_reportedUserId = 0;
	}

	@Override
	protected String getPage() {
		return _PAGE;
	}

	@Override
	protected void setAttributes(HttpServletRequest httpServletRequest) {
		try {
			String message = _getMessage();

			httpServletRequest.setAttribute(
				"liferay-flags:flags:data", _getData(message));

			httpServletRequest.setAttribute(
				"liferay-flags:flags:elementClasses", _elementClasses);
			httpServletRequest.setAttribute(
				"liferay-flags:flags:message", message);
			httpServletRequest.setAttribute(
				"liferay-flags:flags:onlyIcon", !_label);
		}
		catch (Exception exception) {
			_log.error(exception, exception);
		}
	}

	private Map<String, Object> _getData(String message)
		throws PortalException {

		return HashMapBuilder.<String, Object>put(
			"context",
			HashMapBuilder.<String, Object>put(
				"namespace", PortalUtil.getPortletNamespace(PortletKeys.FLAGS)
			).build()
		).put(
			"props",
			() -> {
				HttpServletRequest httpServletRequest = getRequest();

				ThemeDisplay themeDisplay =
					(ThemeDisplay)httpServletRequest.getAttribute(
						WebKeys.THEME_DISPLAY);

				return HashMapBuilder.<String, Object>put(
					"baseData", _getDataJSONObject(themeDisplay)
				).put(
					"captchaURI", FlagsTagUtil.getCaptchaURI(httpServletRequest)
				).put(
					"companyName",
					() -> {
						Company company = themeDisplay.getCompany();

						return company.getName();
					}
				).put(
					"disabled", !_enabled
				).put(
					"forceLogin", !FlagsTagUtil.isFlagsEnabled(themeDisplay)
				).put(
					"message",
					() -> {
						if (Validator.isNotNull(message)) {
							return message;
						}

						return null;
					}
				).put(
					"onlyIcon", !_label
				).put(
					"pathTermsOfUse",
					PortalUtil.getPathMain() + "/portal/terms_of_use"
				).put(
					"reasons",
					FlagsTagUtil.getReasons(
						themeDisplay.getCompanyId(), httpServletRequest)
				).put(
					"signedIn", themeDisplay.isSignedIn()
				).put(
					"uri", FlagsTagUtil.getURI(httpServletRequest)
				).build();
			}
		).build();
	}

	private JSONObject _getDataJSONObject(ThemeDisplay themeDisplay) {
		String namespace = PortalUtil.getPortletNamespace(PortletKeys.FLAGS);

		String contentURL = _contentURL;

		if (Validator.isNull(contentURL)) {
			contentURL = FlagsTagUtil.getCurrentURL(getRequest());
		}

		JSONObject dataJSONObject = JSONUtil.put(
			namespace + "className", _className
		).put(
			namespace + "classPK", _classPK
		).put(
			namespace + "contentTitle", _contentTitle
		).put(
			namespace + "contentURL", contentURL
		).put(
			namespace + "reportedUserId", _reportedUserId
		);

		if (themeDisplay.isSignedIn()) {
			User user = themeDisplay.getUser();

			dataJSONObject.put(
				namespace + "reporterEmailAddress", user.getEmailAddress());
		}

		return dataJSONObject;
	}

	private String _getMessage() {
		ResourceBundle resourceBundle = new AggregateResourceBundle(
			TagResourceBundleUtil.getResourceBundle(pageContext),
			LanguageResources.getResourceBundle(
				PortalUtil.getLocale(getRequest())));

		if (Validator.isNotNull(_message)) {
			return LanguageUtil.get(resourceBundle, _message);
		}

		return LanguageUtil.get(resourceBundle, "report");
	}

	private static final String _PAGE = "/flags/page.jsp";

	private static final Log _log = LogFactoryUtil.getLog(FlagsTag.class);

	private String _className;
	private long _classPK;
	private String _contentTitle;
	private String _contentURL;
	private String _elementClasses;
	private boolean _enabled = true;
	private boolean _label = true;
	private String _message;
	private long _reportedUserId;

}