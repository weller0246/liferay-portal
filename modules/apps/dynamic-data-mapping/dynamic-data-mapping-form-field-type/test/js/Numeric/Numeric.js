/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

import {cleanup, render} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';

import Numeric from '../../../src/main/resources/META-INF/resources/Numeric/Numeric';

const globalLanguageDirection = Liferay.Language.direction;

describe('Field Numeric', () => {
	beforeAll(() => {
		Liferay.Language.direction = {en_US: 'rtl'};
	});

	afterAll(() => {
		Liferay.Language.direction = globalLanguageDirection;
	});

	afterEach(cleanup);

	it('renders the default markup', () => {
		const {container} = render(<Numeric />);

		expect(container).toMatchSnapshot();
	});

	it('has a name', () => {
		const {container} = render(<Numeric name="numericField" />);

		expect(container).toMatchSnapshot();
	});

	it('is not readOnly', () => {
		const {container} = render(<Numeric readOnly={false} />);

		expect(container).toMatchSnapshot();
	});

	it('has a helptext', () => {
		const {container} = render(<Numeric tip="Type something" />);

		expect(container).toMatchSnapshot();
	});

	it('has an id', () => {
		const {container} = render(<Numeric id="ID" />);

		expect(container).toMatchSnapshot();
	});

	it('has a label', () => {
		const {container} = render(<Numeric label="label" />);

		expect(container).toMatchSnapshot();
	});

	it('has a placeholder', () => {
		const {container} = render(<Numeric placeholder="Placeholder" />);

		expect(container).toMatchSnapshot();
	});

	it('is not required', () => {
		const {container} = render(<Numeric required={false} />);

		expect(container).toMatchSnapshot();
	});

	it('renders Label if showLabel is true', () => {
		const {container} = render(
			<Numeric label="Numeric Field" showLabel={true} />
		);

		expect(container).toMatchSnapshot();
	});

	it('has a value', () => {
		const {container} = render(<Numeric value="123" />);

		expect(container).toMatchSnapshot();
	});

	it('has a key', () => {
		const {container} = render(<Numeric key="key" />);

		expect(container).toMatchSnapshot();
	});

	it('fills with an input number', () => {
		let output;
		const {container} = render(
			<Numeric
				onChange={({target: {value}}) => {
					output = value;
				}}
			/>
		);

		const input = container.querySelector('input');

		userEvent.type(input, '2');

		expect(output).toBe('2');
	});

	it('changes the mask type', () => {
		const {container} = render(<Numeric dataType="double" value="22.22" />);

		expect(container.querySelector('input').value).toBe('22.22');
	});

	it('filters the non numeric characters when set to integer', () => {
		let output = '';
		const {container} = render(
			<Numeric
				onChange={({target: {value}}) => {
					output += value;
				}}
			/>
		);

		const input = container.querySelector('input');

		// due to an issue on jest was needed to call type() for each key
		// https://github.com/testing-library/user-event/issues/387

		userEvent.type(input, '3');
		userEvent.type(input, '.');
		userEvent.type(input, '0');

		expect(output).toBe('30');
	});

	it('check field value is the same without decimal symbol when fieldType is integer but it receives a double', () => {
		const {container} = render(<Numeric value="3.8" />);

		const input = container.querySelector('input');

		expect(input.value).toBe('38');
	});

	it('remove decimal symbol from value when changing from decimal to integer when symbol of language is comma', () => {
		const {container} = render(
			<Numeric
				dataType="integer"
				symbols={{decimalSymbol: ','}}
				value="22,82"
			/>
		);

		expect(container.querySelector('input').value).toBe('2282');
	});

	describe('Confirmation Field', () => {
		it('renders the confirmation field with the same data type as the original field', () => {
			render(
				<Numeric
					confirmationValue="22.82"
					dataType="double"
					name="numericField"
					requireConfirmation={true}
				/>
			);

			const confirmationField = document.getElementById(
				'numericFieldconfirmationField'
			);

			expect(confirmationField.value).toBe('22.82');
		});

		it('remove decimal symbol of the confirmation value if the data type is Integer', () => {
			render(
				<Numeric
					confirmationValue="22.82"
					name="numericField"
					requireConfirmation={true}
				/>
			);

			const confirmationField = document.getElementById(
				'numericFieldconfirmationField'
			);

			expect(confirmationField.value).toBe('2282');
		});
	});

	describe('Integer Input Mask toggle', () => {
		it('has an inputMaskFormat', () => {
			const {container} = render(
				<Numeric
					inputMask={true}
					inputMaskFormat="+99 (99) 9999-9999"
					name="numericField"
					value="123456789012"
				/>
			);

			expect(container).toMatchSnapshot();
		});

		it('applies mask to value', () => {
			const {container} = render(
				<Numeric
					inputMask={true}
					inputMaskFormat="+99 (99) 9999-9999"
					value="123456789012"
				/>
			);

			const input = container.querySelector('input');

			expect(input.value).toBe('+12 (34) 5678-9012');
		});

		it('applies mask to predefined value', () => {
			const {container} = render(
				<Numeric
					inputMask={true}
					inputMaskFormat="+99 (99) 9999-9999"
					predefinedValue="123456789012"
				/>
			);

			const input = container.querySelector('input');

			expect(input.value).toBe('+12 (34) 5678-9012');
		});

		it('truncates values over mask digit limit', () => {
			const {container} = render(
				<Numeric
					inputMask={true}
					inputMaskFormat="+99 (099) 9999-9999"
					value="12345678901234"
				/>
			);

			const input = container.querySelector('input');

			expect(input.value).toBe('+12 (345) 6789-0123');
		});

		it('ignores optional digits whenever input is less than mandatory', () => {
			const {container} = render(
				<Numeric
					inputMask={true}
					inputMaskFormat="+09 (099) 9999-9999"
					value="12345"
				/>
			);

			const input = container.querySelector('input');

			expect(input.value).toBe('+1 (23) 45');
		});

		it('sends unmasked value though onChange event', () => {
			let output = null;
			const {container} = render(
				<Numeric
					inputMask={true}
					inputMaskFormat="+99 (99) 9999-9999"
					onChange={({target: {value}}) => {
						output = value;
					}}
				/>
			);

			const input = container.querySelector('input');

			userEvent.type(input, '+55 (81) 2121-6000');

			expect(output).toBe('558121216000');
		});

		it('limits predefined value size according to the mask', () => {
			const {container} = render(
				<Numeric
					inputMask={true}
					inputMaskFormat="99-99"
					name="LPS-134259"
					predefinedValue="12345"
				/>
			);

			const input = container.querySelector('input[name="LPS-134259"]');

			expect(input.value).toBe('1234');
		});

		/**
		 * This test was skipped due to an issue on userEvent.type() that not
		 * allows simulate backspace key pressing (with the current
		 * @testing-library/use-event)
		 */
		xit('it allows to delete non numeric characters from mask', () => {
			const {container} = render(
				<Numeric
					inputMask={true}
					inputMaskFormat="99-99"
					onChange={() => {}}
					predefinedValue="12"
				/>
			);

			const input = container.querySelector('input');

			userEvent.click(input);
			userEvent.type(input, '{backspace}');

			expect(input.value).toBe('1');
		});
	});

	describe('Decimal Input Mask toggle', () => {
		it('renders a suffix', () => {
			const {container} = render(
				<Numeric
					append="$"
					appendType="suffix"
					dataType="double"
					inputMask={true}
					name="numericField"
					value="123"
				/>
			);

			const input = container.querySelector('input');

			expect(input.value).toBe('123$');
		});

		it('renders a prefix', () => {
			const {container} = render(
				<Numeric
					append="$"
					appendType="prefix"
					dataType="double"
					inputMask={true}
					name="numericField"
					value="123"
				/>
			);

			const input = container.querySelector('input');

			expect(input.value).toBe('$123');
		});

		it('renders the thousand separator', () => {
			const {container} = render(
				<Numeric
					dataType="double"
					inputMask={true}
					name="numericField"
					symbols={{decimalSymbol: '.', thousandsSeparator: ','}}
					value="1234"
				/>
			);

			const input = container.querySelector('input');

			expect(input.value).toBe('1,234');
		});

		it('allows user to input a decimal separator', () => {
			const onChange = jest.fn();
			const {container} = render(
				<Numeric
					dataType="double"
					inputMask={true}
					name="numericField"
					onChange={onChange}
					symbols={{decimalSymbol: ','}}
				/>
			);

			const input = container.querySelector('input');

			userEvent.type(input, '1,234');

			expect(onChange.mock.calls[4][0].target.value).toBe('1,23');
		});
	});
});
