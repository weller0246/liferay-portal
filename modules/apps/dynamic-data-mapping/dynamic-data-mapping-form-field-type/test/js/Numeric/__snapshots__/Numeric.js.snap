// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field Numeric Integer Input Mask toggle has an inputMaskFormat 1`] = `
<div>
  <div
    aria-labelledby="numericField"
    class="form-group hide"
    data-field-name="numericField"
    tabindex="0"
  >
    <input
      class="form-control"
      dir="rtl"
      lang="en_US"
      name="numericField_masked"
      placeholder="+__ (__) ____-____"
      type="text"
      value="+12 (34) 5678-9012"
    />
    <input
      name="numericField"
      type="hidden"
      value="123456789012"
    />
    <input
      name="numericField_edited"
      type="hidden"
      value="false"
    />
    
  </div>
</div>
`;

exports[`Field Numeric has a helptext 1`] = `
<div>
  <div
    class="form-group hide"
    tabindex="0"
  >
    <input
      class="form-control"
      dir="rtl"
      lang="en_US"
      name="undefined"
      type="text"
      value=""
    />
    <input
      name="undefined_edited"
      type="hidden"
      value="false"
    />
    <span
      aria-hidden="true"
      class="form-text"
    >
      Type something
    </span>
    <span
      class="sr-only"
    >
      Type something
      <br />
    </span>
  </div>
</div>
`;

exports[`Field Numeric has a key 1`] = `
<div>
  <div
    class="form-group hide"
    tabindex="0"
  >
    <input
      class="form-control"
      dir="rtl"
      lang="en_US"
      name="undefined"
      type="text"
      value=""
    />
    <input
      name="undefined_edited"
      type="hidden"
      value="false"
    />
    
  </div>
</div>
`;

exports[`Field Numeric has a label 1`] = `
<div>
  <div
    class="form-group hide"
  >
    <label
      class="ddm-label"
      tabindex="0"
    >
      label
    </label>
    <input
      class="form-control"
      dir="rtl"
      lang="en_US"
      name="undefined"
      type="text"
      value=""
    />
    <input
      name="undefined_edited"
      type="hidden"
      value="false"
    />
    
  </div>
</div>
`;

exports[`Field Numeric has a name 1`] = `
<div>
  <div
    aria-labelledby="numericField"
    class="form-group hide"
    data-field-name="numericField"
    tabindex="0"
  >
    <input
      class="form-control"
      dir="rtl"
      lang="en_US"
      name="numericField"
      type="text"
      value=""
    />
    <input
      name="numericField_edited"
      type="hidden"
      value="false"
    />
    
  </div>
</div>
`;

exports[`Field Numeric has a placeholder 1`] = `
<div>
  <div
    class="form-group hide"
    tabindex="0"
  >
    <input
      class="form-control"
      dir="rtl"
      lang="en_US"
      name="undefined"
      placeholder="Placeholder"
      type="text"
      value=""
    />
    <input
      name="undefined_edited"
      type="hidden"
      value="false"
    />
    
  </div>
</div>
`;

exports[`Field Numeric has a value 1`] = `
<div>
  <div
    class="form-group hide"
    tabindex="0"
  >
    <input
      class="form-control"
      dir="rtl"
      lang="en_US"
      name="undefined"
      type="text"
      value="123"
    />
    <input
      name="undefined_edited"
      type="hidden"
      value="false"
    />
    
  </div>
</div>
`;

exports[`Field Numeric has an id 1`] = `
<div>
  <div
    aria-labelledby="ID"
    class="form-group hide"
    tabindex="0"
  >
    <input
      class="form-control"
      dir="rtl"
      id="ID"
      lang="en_US"
      name="undefined"
      type="text"
      value=""
    />
    <input
      name="undefined_edited"
      type="hidden"
      value="false"
    />
    
  </div>
</div>
`;

exports[`Field Numeric is not readOnly 1`] = `
<div>
  <div
    class="form-group hide"
    tabindex="0"
  >
    <input
      class="form-control"
      dir="rtl"
      lang="en_US"
      name="undefined"
      type="text"
      value=""
    />
    <input
      name="undefined_edited"
      type="hidden"
      value="false"
    />
    
  </div>
</div>
`;

exports[`Field Numeric is not required 1`] = `
<div>
  <div
    class="form-group hide"
    tabindex="0"
  >
    <input
      class="form-control"
      dir="rtl"
      lang="en_US"
      name="undefined"
      type="text"
      value=""
    />
    <input
      name="undefined_edited"
      type="hidden"
      value="false"
    />
    
  </div>
</div>
`;

exports[`Field Numeric renders Label if showLabel is true 1`] = `
<div>
  <div
    class="form-group hide"
  >
    <label
      class="ddm-label"
      tabindex="0"
    >
      Numeric Field
    </label>
    <input
      class="form-control"
      dir="rtl"
      lang="en_US"
      name="undefined"
      type="text"
      value=""
    />
    <input
      name="undefined_edited"
      type="hidden"
      value="false"
    />
    
  </div>
</div>
`;

exports[`Field Numeric renders the default markup 1`] = `
<div>
  <div
    class="form-group hide"
    tabindex="0"
  >
    <input
      class="form-control"
      dir="rtl"
      lang="en_US"
      name="undefined"
      type="text"
      value=""
    />
    <input
      name="undefined_edited"
      type="hidden"
      value="false"
    />
    
  </div>
</div>
`;
